### PR TITLE
Db model integration 

### PR DESCRIPTION
**Pull request by chmette**
@AdrianValdes sorry that its so big.

done: 
- build Schemas and models 
- created all controllers for db communication
- testing
-  integrated lint back ins PR file 
- little bug fixes

open:
- image integration avatar as well as records
- testing again

### CODE REVIEW

File: src/controller/recordController.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//   const { id } = req.params;
+//   try {
+//     const UpdatedRecord = await Record.findByIdAndUpdate(id, req.body, {
+//       new: true,
+//     });
+//     res.json(UpdatedRecord);
+//   } catch (err) {
+//     next(err);
+//   }
+// };
```
Comment (AdrianValdes) [inline src/controller/recordController.js]: Could we avoid pushing commented code to main? You can have things to avoid something breaking, but commented code is never good.

**Comment by AdrianValdes** [comment]
@chmette
> refactor: cean up code

 :D

File: src/controller/recordController.js (hunk 1 above)
Comment (chmette) [reply src/controller/recordController.js]: removed

**Comment by AdrianValdes** [comment]
You haven't rename the images yet, have you? I am asking because sometimes Github gets crazy with file changes

File: src/controller/recordController.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// for testing
+const getRecord = async (req, res, next) => {
+  const { id } = req.params; // tbd
+  try {
+    const record = await Record.findById(id);
+    res.json(record);
+  } catch (err) {
+    next(err);
+  }
+};
```
Comment (AdrianValdes) [inline src/controller/recordController.js]: Are we using this function? I know it says for testing but if we are not using it we shouldn't have it there. It is not good to have code it is not being used at all. You can always save it somewhere, or having in another branch, but it is not good practice to merge it into main.

File: src/controller/userController.js [hunk 3]
```diff
@@ -1,11 +1,79 @@
...
+
+// for testing
+const getUsers = async (req, res, next) => {
+  try {
+    const allUsers = await User.find();
+    res.json(allUsers);
+  } catch (err) {
+    next(err);
+  }
+};
```
Comment (AdrianValdes) [inline src/controller/userController.js]: I actually wanted to ask about if we are using this function or not. I think what I said apply for both

File: src/controller/userController.js [hunk 4]
```diff
@@ -1,11 +1,79 @@
...
+    const error = new Error(
+      `User with ${email} does not exist. Please signup.`
+    );
+    error.status = 400;
+    next(error);
+  }
+};
+
+const getUser = async (req, res, next) => {
+  const { id } = req.params;
```
Comment (AdrianValdes) [inline src/controller/userController.js]: are we getting the user by id? Or by email?

File: src/controller/userController.js (hunk 3 above)
Comment (chmette) [reply src/controller/userController.js]: for testing reasons I will leave it there if you don't mind. We need to write the script to upload like 20 different files at once , and if this works fine i will take it directly out ok?

File: src/controller/userController.js (hunk 4 above)
Comment (chmette) [reply src/controller/userController.js]: I build it for the case, that an already logged in user, wants to see his profile - we don't need to get the user details via log in then, (especially not if we use a token). Wrong thought? What you think?

File: src/controller/recordController.js (hunk 2 above)
Comment (chmette) [reply src/controller/recordController.js]: will delete this one

File: src/controller/userController.js (hunk 3 above)
Comment (AdrianValdes) [reply src/controller/userController.js]: I understand, you want to use it, but main it is branch that it shouldn't be used for that. Now we would have a 10 lines function that you could easily make and we are not using, you can create a branch for that and leave there if you don't want to delete it. We should only merge to main production-code, not testing code.

Comment (AdrianValdes) [reply src/controller/userController.js]: When we want to write a `placeholder` we can do it. Something that we need to write because if not the code would be too big. Like a response to a `catch` to the error handlers you have for now, they are not ideal, but they have to be there to not break the code. But this case is different.

Comment (AdrianValdes) [reply src/controller/userController.js]: Always think the code has to follow some logic, at least the code in `main`, right now the rest of the team is seeing a function that is declared but never used. So if that is in main I will always wonder what we are not using it. If something is not being used, shouldn't be in main.

Comment (AdrianValdes) [reply src/controller/userController.js]: Right now the code it is a little hard to follow because there is a few functions declared and exported and are not being used. If you think so, actually the `linter` doesn't allow unused vars at all, I am wondering how you can have them there. Ah, because you are exporting them, right

File: src/controller/userController.js (hunk 4 above)
Comment (AdrianValdes) [reply src/controller/userController.js]: Hmm, I thought about that, it is not crazy but we have something better I think. With what we have once the user logs in, we are going to save its info in context so we don't to make more api calls.

File: src/controller/userController.js (hunk 3 above)
Comment (chmette) [reply src/controller/userController.js]: its deleted

File: src/controller/userController.js (hunk 4 above)
Comment (AdrianValdes) [reply src/controller/userController.js]: So basically the email and password is our kind of firewall to get the user. I am thinking if we need the user id to get the records, but even in that case we will send back that info once the user logs in.

File: src/controller/userController.js [hunk 5]
```diff
@@ -1,11 +1,68 @@
...
+  try {
+    const user = await User.findById(id);
+    res.json(user);
+  } catch (err) {
+    next(err);
+  }
+};
```
Comment (AdrianValdes) [inline src/controller/userController.js]: So in the end we don't need this one, right? The getUser with the id

File: src/routes/usersRouter.js [hunk 6]
```diff
@@ -1,9 +1,14 @@
...
+const {
+  updateUser,
+  deleteUser,
+  getUser,
+} = require('../controller/userController');
 
-//users == base path
-router.route('/:id').put(updateUser).delete(deleteUser);
+// users == base path
+router.route('/:id').put(updateUser).delete(deleteUser).get(getUser);
```
Comment (AdrianValdes) [inline src/routes/usersRouter.js]: Could we use `.patch` to update? It is more declarative because we would have a distinction in our code between creating and updating.

Comment (chmette) [reply src/routes/usersRouter.js]: why do you want to use patch?

File: src/controller/userController.js (hunk 5 above)
Comment (chmette) [reply src/controller/userController.js]: we can add it in case we do , right. so ill take it finally out now.

File: src/models/Record.js [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+const { Schema, model } = mongoose;
+
+const RecordSchema = new Schema(
+  {
+    cover: { type: String, required: true, default: 'xyz' },
+    title: { type: String, required: true },
+    artist: { type: String, required: false },
+    year: { type: Number, required: false },
+  },
```
Comment (AdrianValdes) [inline src/models/Record.js]: Could we make these two required too, those are pretty important for a record, and it makes the work easier when we display it in the front-end. We are going to have something checking probably but in this case makes sense the artist and the year are required anyways, normally those are easy to find. In our case it will be fake of course, it just for consinstency.

Comment (chmette) [reply src/models/Record.js]: sure

File: src/routes/usersRouter.js (hunk 6 above)
Comment (AdrianValdes) [reply src/routes/usersRouter.js]: because then we will have patch for updating and post for creating. Remember to google before asking

PATCH
The HTTP PATCH method is defined in RFC 5789 as an extension to the earlier mentioned HTTP RFC. While PUT is used to replace an existing resource, PATCH is used to apply partial modifications to a resource.

Comment (AdrianValdes) [reply src/routes/usersRouter.js]: Also when Robert saw that Vasilis used post to update he made a comment on that.

File: src/models/Record.js (hunk 7 above)
Comment (AdrianValdes) [reply src/models/Record.js]: thanks

File: src/routes/usersRouter.js (hunk 6 above)
Comment (chmette) [reply src/routes/usersRouter.js]: i googles before... just wanted to know your reasons. I am aware of the difference between put and patch.
we are not using post, i used put. and both can be used for updates. there fore i was asking for your reasons.

Comment (AdrianValdes) [reply src/routes/usersRouter.js]: Oh, but those ones I said it here:

>  It is more declarative because we would have a distinction in our code between creating and updating.

Comment (chmette) [reply src/routes/usersRouter.js]: if we use patch : _"This means that you are only required to send the data that you want to update, and it won’t affect or change anything else. So if you want to update the first name on a database, you will only be required to send the first parameter; the first name."_

put: _"PUT is a method of modifying resource where the client sends data that updates the entire resource. It is used to set an entity’s information completely. PUT is similar to POST in that it can create resources, but it does so when there is a defined URI. PUT overwrites the entire entity if it already exists, and creates a new resource if it doesn’t exist.

For example, when you want to change the first name of a person in a database, you need to send the entire resource when making a PUT request."_

Comment (chmette) [reply src/routes/usersRouter.js]: i dont mind using patch i just want to understand why you prefer it .

Comment (AdrianValdes) [reply src/routes/usersRouter.js]: PUT can be for both, PATCH it is only update, so there is no wondering, if an engineer sees the code, knows immediately   that that is an update.

Comment (chmette) [reply src/routes/usersRouter.js]: changed.

Comment (AdrianValdes) [reply src/routes/usersRouter.js]: > For example, when you want to change the first name of a person in a database, you need to send the entire resource when making a PUT request."_

Right and we don't want that.

Comment (AdrianValdes) [reply src/routes/usersRouter.js]: Could you see how it could be inefficient to send the whole item if you only need to change a field? And also how it makes the code more declarative and therefore helps to have specialized methods if we can? Or it doesn't make too much sense?

Comment (chmette) [reply src/routes/usersRouter.js]: it makes sense.

**Comment by AdrianValdes** [comment]
Remember to re-request the review

**Review by AdrianValdes — APPROVED** [review]
Ok, looks good. Thanks for the hard work! Into main!
